File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.js
function solution (numbers, target) {  
    let start = numbers[0];
    let cnt = 0;
    let sum = 0;
    
    function dfs (sum, number, idx) {
        if (idx < numbers.length) {
            dfs (sum + number, numbers[idx + 1], idx + 1);
            dfs (sum - number, numbers[idx + 1], idx + 1);
        }
        
        if (sum === target && idx >= numbers.length) cnt++;
    }  
    
    dfs(sum, start, 0);
    
    return cnt;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def dfs(depth, target, cur_sum, n, numbers):
    if depth == n:
        return 1 if cur_sum == target else 0
    
    return dfs(depth + 1, target, cur_sum + numbers[depth], n, numbers) + dfs(depth + 1, target, cur_sum - numbers[depth], n, numbers)

def solution(numbers, target):
    n = len(numbers)
    
    return dfs(0, target, 0, n, numbers)

